### PR TITLE
Fix for support on different Doctrine ORM versions

### LINKED ISSUE #413: Faker/ORM/Doctrine not working with latest doctrine implementations
### Summary
The current version of faker uses the following classes of Doctrine

- `Doctrine\Common\Persistence\ObjectManager`
- `Doctrine\Common\Persistence\Mapping\ClassMetadata`

As seen here:
https://github.com/FakerPHP/Faker/tree/main/src/Faker/ORM/Doctrine

`doctrine/persistence` with version `1.3` moved the classes around.
With version `2.0` these aliases were gone, this is when faker will stop working.

We have two options here:
1. Release a fix adding a simple alias (same as doctrine/persistence https://github.com/doctrine/persistence/blob/1.3.x/lib/Doctrine/Common/Persistence/ObjectManager.php)
2. Add a `conflict` in `composer.json` to avoid installing with doctrine/persistence below ^2.0 and use the new classes.

### Versions

|                | Version |
|:---------------|:--------|
| PHP            | 7.4.x   |
| `fakerphp/faker`  | latest   |

### Current quickfix
To quickly work around this issue I added these two lines into my `tests/bootstrap.php`

```
class_alias(\Doctrine\Persistence\ObjectManager::class, 'Doctrine\Common\Persistence\ObjectManager');
class_alias(\Doctrine\ORM\Mapping\ClassMetadata::class, 'Doctrine\Common\Persistence\Mapping\ClassMetadata');
```

### PR DESCRIPTION
**Pull request by icanhazstring**
Add conflict to doctrine/persistence below 1.4 and use correct persistence classes

### What is the reason for this PR?

Fix an issue with `doctrine/persistnce`
- [x] Fixed an issue (resolve #413 )

### Author's checklist

- [x] Follow the [Contribution Guide](https://github.com/FakerPHP/Faker/blob/main/.github/CONTRIBUTING.md)

### Summary of changes

- ~Add a conflict into `composer.json` for `doctrine/persistence` below version 1.4~
- Add suggest into `composer.json` for `doctrine/orm`
- ~Add `doctrine/orm` as dev requirement to avoid static analysis baseline entries~
- ~Use new classes of `doctrine/orm` and `doctrine/persistence` for `Faker\ORM\Doctrine` namespace~
- Add `class_alias` for `ClassMetadata` and `ObjectManager`

### Review checklist

- [x] All checks have passed
- [ ] Changes are approved by maintainer

### CODE REVIEW

**Comment by pimjansen** [comment]
This is a BC which we do not allow for v1. Can we ensure that both are supported here?

**Comment by icanhazstring** [comment]
@pimjansen I can try and do the same doctrine did and add a `class_alias`
So we could then remove it with v2.

**Comment by pimjansen** [comment]
> @pimjansen I can try and do the same doctrine did and add a `class_alias`
> 
> So we could then remove it with v2.

Correct. For v2 we will remove it from core and can be a community driven extension instead.

**Comment by icanhazstring** [comment]
Apparently, psalm is failing in CI but not on my machine using the same PHP version.

```
vendor/bin/psalm
Scanning files...
Analyzing files...

░
------------------------------
No errors found!
------------------------------

Checks took 0.71 seconds and used 234.707MB of memory
Psalm was unable to infer types in the codebase

```

**Comment by fdiedler** [comment]
Hope it will be fixed soon, thanks :)

**Comment by pimjansen** [comment]
@icanhazstring the pipeline is failing, can you review this?

**Comment by icanhazstring** [comment]
> Apparently, psalm is failing in CI but not on my machine using the same PHP version.
> 
> ```
> vendor/bin/psalm
> Scanning files...
> Analyzing files...
> 
> ░
> ------------------------------
> No errors found!
> ------------------------------
> 
> Checks took 0.71 seconds and used 234.707MB of memory
> Psalm was unable to infer types in the codebase
> 
> ```

@pimjansen see this comment.

**Comment by pimjansen** [comment]
No idea but ci is leading so needs to be green

File: psalm.baseline.xml [hunk 1]
```diff
@@ -1,235 +1,2 @@
...
-    </InvalidReturnStatement>
-    <InvalidReturnType occurrences="1">
-      <code>string</code>
-    </InvalidReturnType>
-  </file>
-  <file src="src/Faker/Generator.php">
-    <InvalidThrow occurrences="1">
-      <code>ContainerExceptionInterface</code>
-    </InvalidThrow>
-  </file>
```
Comment (krsriq) [inline psalm.baseline.xml:15]: Should these entries really be removed?

File: psalm.baseline.xml [hunk 2]
```diff
@@ -1,235 +1,2 @@
...
-    <UndefinedDocblockClass occurrences="1">
-      <code>DateTime</code>
-    </UndefinedDocblockClass>
-  </file>
-  <file src="src/Faker/Provider/sl_SI/Person.php">
-    <NonStaticSelfCall occurrences="2">
-      <code>static::lastName()</code>
-      <code>static::lastName()</code>
-    </NonStaticSelfCall>
-  </file>
```
Comment (krsriq) [inline psalm.baseline.xml:234]: I think these should stay as well.

File: src/Faker/ORM/Doctrine/EntityPopulator.php [hunk 3]
```diff
@@ -226,20 +226,17 @@ private function callMethods($obj, $insertedEntities)
...
                 ->select(sprintf('e.%s', $column))
                 ->getQuery()
                 ->execute();
         $ids = array_map('current', $result->toArray());
 
-        $id = null;
-
         do {
             $id = mt_rand();
         } while (in_array($id, $ids, false));
```
Comment (krsriq) [inline src/Faker/ORM/Doctrine/EntityPopulator.php:242]: `$id` can not be null here since the `do {` loop executes at least once, right?

**Comment by krsriq** [comment]
@icanhazstring Running psalm locally is red for me too, works with this psalm.baseline.xml: https://gist.github.com/krsriq/cc1c3ceea2886aa4b93a2b1825a32e7f

**Comment by icanhazstring** [comment]
@krsriq then there is something weird with the local setup 🤷‍♂️

Guess I'll just add your baseline then.

**Comment by icanhazstring** [comment]
After all these years.
Now it works.

File: psalm.baseline.xml (hunk 2 above)
Comment (icanhazstring) [reply psalm.baseline.xml]: Found a way to correctly run psalm now, so reverted to main + update

**Comment by bram-pkg** [comment]
@icanhazstring is this ready now?

**Comment by icanhazstring** [comment]
@bram-pkg checked it again against my problem. Fixed for me 👍 
LGTM

**Review by pimjansen — APPROVED** [review]
LGTM, also test on 1.3 and 2.x went fine

**Comment by crynobone** [comment]
This PR caused an issue on projects that don't utilize `FakerPHP\Doctrine` 

```
Warning: Class "Doctrine\Persistence\Mapping\ClassMetadata" not found in /Projects/laravel/nova/orion/vendor/fakerphp/faker/src/Faker/ORM/Doctrine/backward-compatibility.php on line 6

Warning: Class "Doctrine\Persistence\ObjectManager" not found in /Projects/laravel/nova/orion/vendor/fakerphp/faker/src/Faker/ORM/Doctrine/backward-compatibility.php on line 10
```

**Comment by icanhazstring** [comment]
It was added to `require-dev` because it is used for development.

But the other case you are correct. Will create a quick fix for that.

**Comment by pimjansen** [comment]
@crynobone fix has been applied in main

**Pull request merged**